### PR TITLE
[RLlib] Issue 40347: Fix saving and loading algos by loading stored learner state.

### LINKED ISSUE #40347: Issue with /rllib/rllib-saving-and-loading-algos-and-policies
Good afternoon! 

I have trained PPO algorithm with environment MountainCarContinuous-v0. 

            ray_module = importlib.import_module(configuration["algo_configuration"]["module"])
            ray_config_class = getattr(ray_module, configuration["algo_configuration"]["class"])
            algo_config = (
                ray_config_class()
                .environment("MountainCarContinuous-v0", env_config=configuration["env_configuration"])
                .framework("torch")
            )
            for configuration_branch in ("resources", "rollouts", "training",
                                        "optimizer", "evaluation", "exploration"):
                if configuration.get(f"{configuration_branch}_configuration"):
                    algo_config = getattr(algo_config, configuration_branch)(
                        **configuration.get(f"{configuration_branch}_configuration")
                    )
            algo = algo_config.build()
            result = algo.train()

After some time of training I receive needed reward metrics 
episode_reward_max: MAX = -0.067 (step=832) Latest= -0.067 (step=1219)
image
I make checkpoints every 500 steps. 
            ...
                checkpoint_result = algo.save(checkpoint_path)
                if isinstance(checkpoint_result, (str, os.PathLike)):
                    saved_checkpoint = checkpoint_result
                elif hasattr(checkpoint_result, "checkpoint"):
                    saved_checkpoint = checkpoint_result.checkpoint.path

After that I want to continue training from checkpoint (step 999): 

        if arguments.continue_run:
            logging.info("Contuning run %s", arguments.continue_run)
            checkpoint_path = "/ray/iteration_999/checkpoints/oct_golod_BipedalWalker-v3_20231014105544_999"
            algo = Algorithm.from_checkpoint(checkpoint_path)
...
      result = algo.train()

And I receive metrics in the beginning of the training as the algorithm was not trained before. 
image

Please advise me, how can I solve this problem and restore from checkpoint without loosing training experience. 

Thank you in advance!
Kind regards, Alexander!

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

We had already added a possibility to store the learner state when the new stack is enabled, i.e. `_enable_new_api_stack=True`, but have not supported to restore this state when training on further.  This PR closes this gap by loading a stored learner state when the user calls `Algorithm.from_checkpoint()`.

For the `DreamerV3` algorithm it needs an additional update of the `RLModule` in the `EnvRunner`  as after the `setup()` is called, the `__setstate__()`  is loading the state from checkpoint into the `Learner.module`. 

## Related issue number

Closes #40347 #41290 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simonsays1980** [comment]
Output from running the script provided in #41290 (the part after reloading) using the modifications in this PR:

```2023-12-27 19:19:54,789 INFO trainable.py:188 -- Trainable.setup took 21.533 seconds. If your trainable is slow to initialize, consider setting reuse_actors=True to reduce actor creation overheads.
2023-12-27 19:19:54,790 WARNING util.py:68 -- Install gputil for GPU system monitoring.
epoch=0 reward_max=357.000000 reward_mean=252.666667
epoch=1 reward_max=357.000000 reward_mean=357.000000
epoch=2 reward_max=335.000000 reward_mean=335.000000
epoch=3 reward_max=335.000000 reward_mean=335.000000
epoch=4 reward_max=335.000000 reward_mean=335.000000
epoch=5 reward_max=335.000000 reward_mean=335.000000
epoch=6 reward_max=335.000000 reward_mean=335.000000
epoch=7 reward_max=335.000000 reward_mean=335.000000
epoch=8 reward_max=335.000000 reward_mean=335.000000
epoch=9 reward_max=500.000000 reward_mean=500.000000
```

**Comment by simonsays1980** [comment]
Maybe similar to PR #39209

File: rllib/algorithms/algorithm.py [hunk 1]
```diff
@@ -2815,6 +2814,15 @@ def _setup_eval_worker(w):
                     "data found in state!"
                 )
 
+        if self.config._enable_new_api_stack:
+            if "learner_state_dir" in state:
+                self.learner_group.load_state(state["learner_state_dir"])
+            else:
+                logger.warning(
+                    "`_enable_new_api_stack` is `True`, but no learner state "
+                    "directory found."
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:2823]:
```suggestion
                    "You configured `_enable_new_api_stack=True`, but no `learner_state_dir` "
                    "key could be found in the state dict!"
```

File: rllib/algorithms/dreamerv3/dreamerv3.py [hunk 2]
```diff
@@ -698,3 +698,27 @@ def training_ratio(self) -> float:
     @staticmethod
     def _reduce_results(results: List[Dict[str, Any]]):
         return tree.map_structure(lambda *s: np.mean(s, axis=0), *results)
+
+    @PublicAPI
```
Comment (sven1977) [inline rllib/algorithms/dreamerv3/dreamerv3.py:702]:
```suggestion

    # TODO (sven): Remove this once DreamerV3 is on the new SingleAgentEnvRunner.
    @PublicAPI
```

**Review by sven1977 — APPROVED** [review]
LGTM. Thanks a ton for this fix @simonsays1980 !

File: rllib/algorithms/algorithm.py [hunk 3]
```diff
@@ -2815,6 +2814,15 @@ def _setup_eval_worker(w):
                     "data found in state!"
                 )
 
+        if self.config._enable_new_api_stack:
+            if "learner_state_dir" in state:
+                self.learner_group.load_state(state["learner_state_dir"])
+            else:
+                logger.warning(
+                    "You configured `_enable_new_api_stack=True`, but no `learner_state_dir` "
+                    "key could be found in the state dict!"
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:2823]:
```suggestion
                    "You configured `_enable_new_api_stack=True`, but no 
                    "`learner_state_dir` key could be found in the state dict!"
```

**Pull request merged**